src/module.c

```diff
@@ -3048,6 +3051,264 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
+ * allowed.
+ */
+int RM_StreamAdd(RedisModuleKey *key, int flags, RedisModuleStreamID *id, RedisModuleString **argv, long numfields) {
+    /* Validate args */
+    if (flags & ~(REDISMODULE_STREAM_AUTOID |
+                  REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR; /* unknown flags */
+    if (!(key->mode & REDISMODULE_WRITE))
+        return REDISMODULE_ERR;
+    if (key->value && key->value->type != OBJ_STREAM)
+        return REDISMODULE_ERR;
+    if (key->value == NULL && (flags & REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR;
+    if (!(flags & REDISMODULE_STREAM_AUTOID) &&
+        (id == NULL || (id->ms == 0 && id->seq == 0)))
+        return REDISMODULE_ERR;
+
+    /* Create key if necessery */
+    if (key->value == NULL) {
+        moduleCreateEmptyKey(key, REDISMODULE_KEYTYPE_STREAM);
+    }
+
+    stream *s = key->value->ptr;
+    streamID added_id;
+    streamID use_id;
+    streamID *use_id_ptr = NULL;
+    if (!(flags & REDISMODULE_STREAM_AUTOID)) {
+        use_id.ms = id->ms;
+        use_id.seq = id->seq;
+        use_id_ptr = &use_id;
+    }
+    if (streamAppendItem(s, argv, numfields, &added_id, use_id_ptr) == C_ERR) {
+        /* Current last ID is greater than or equal to 'use_id' */
+        return REDISMODULE_ERR;
+    }
+
+    if (id != NULL) {
+        id->ms = added_id.ms;
+        id->seq = added_id.seq;
+    }
+
+    return REDISMODULE_OK;
+}
+
+/* Sets up a stream iterator.
+ *
+ * - `key`: The stream key opened for reading using RedisModule_OpenKey().
+ * - `flags`:
+ *   - `REDISMODULE_STREAM_EXCLUSIVE`: Don't include minid and maxid in the
+ *     iterated range.
+ *   - `REDISMODULE_STREAM_REVERSE`: Iterate in reverse order, starting from
+ *     maxid.
+ * - `minid`: Lower bound of the iterated range. May be REDISMODULE_STREAM_MIN.
+ * - `maxid`: Upper bound of the iterated range. May be REDISMODULE_STREAM_MAX.
+ *
+ * Returns REDISMODULE_OK on success and REDISMODULE_ERR if the key doesn't
+ * refer to a stream or if invalid arguments were given.
+ *
+ * The stream items are retrieved using RedisModule_StreamIteratorNext(). The
+ * iterator must be freed by callling RedisModule_StreamIteratorStop().
+ */
+int RM_StreamIteratorStart(RedisModuleKey *key, int flags, RedisModuleStreamID *minid, RedisModuleStreamID *maxid) {
+    /* check args */
+    if (flags & ~(REDISMODULE_STREAM_EXCLUSIVE |
+                  REDISMODULE_STREAM_REVERSE)) {
+        return REDISMODULE_ERR; /* invalid flags */
+    }
+    if (!key || !key->value || key->value->type != OBJ_STREAM) {
+        return REDISMODULE_ERR; /* not a stream */
+    }
+
+    /* define range */
+    streamID min_id = minid ? (streamID){minid->ms, minid->seq}
+                            : (streamID){0, 0};
+    streamID max_id = maxid ? (streamID){maxid->ms, maxid->seq}
+                            : (streamID){UINT64_MAX, UINT64_MAX};
```

Comment:
perhaps you want to use the MIN/MAX constants here

Reply:
Yes, the user can use it like
```C
    RedisModuleStreamID startid = REDISMODULE_STREAM_MIN;
    RedisModuleStreamID endid = REDISMODULE_STREAM_MAX;
    if (RedisModule_StreamIteratorStart(key, 0, &startid, &endid);
```

... but since they are pointers, they can also be NULL. With NULL we also use MIN and MAX.

In the other comment, I suggested we don't let it be pointers. Then they can't be null and we don't need this logic.

Reply:
generally speaking we prefer to pass structs by ref and not by value.. i think it's ok as-is